smt/pkg/smt/smt_state_reader.go

```diff
@@ -0,0 +1,245 @@
+func (s *SMT) GetAccountBalance(address libcommon.Address) (*uint256.Int, error) {
+	balance := uint256.NewInt(0)
+
+	valueInBytes, err := s.getValueInBytes(utils.KEY_BALANCE, address)
+	if err != nil {
+		log.Error("error getting balance", "error", err)
+		return nil, err
+	}
+	balance.SetBytes(valueInBytes)
+
+	return balance, nil
+}
+
+// GetAccountNonce returns the nonce of an account from the SMT
+func (s *SMT) GetAccountNonce(address libcommon.Address) (*uint256.Int, error) {
+	nonce := uint256.NewInt(0)
+
+	valueInBytes, err := s.getValueInBytes(utils.KEY_NONCE, address)
+	if err != nil {
+		log.Error("error getting nonce", "error", err)
+		return nil, err
+	}
+	nonce.SetBytes(valueInBytes)
+
+	return nonce, nil
+}
+
+// GetAccountCodeHash returns the code hash of an account from the SMT
+func (s *SMT) GetAccountCodeHash(address libcommon.Address) (libcommon.Hash, error) {
+	codeHash := libcommon.Hash{}
+
+	valueInBytes, err := s.getValueInBytes(utils.SC_CODE, address)
+	if err != nil {
+		log.Error("error getting codehash", "error", err)
+		return libcommon.Hash{}, err
+	}
+	codeHash.SetBytes(valueInBytes)
+
+	return codeHash, nil
+}
+
+// GetAccountStorageRoot returns the StorageRoot of an account from the SMT (not implemented for SMT)
+func (s *SMT) GetAccountStorageRoot(address libcommon.Address) (libcommon.Hash, error) {
+	storageTrie := trie.New(libcommon.Hash{})
+
+	storageMap, err := s.getStorageMap(address)
+	if err != nil {
+		return libcommon.Hash{}, err
+	}
+
+	for key, value := range storageMap {
+		storageTrie.Update(key[:], value[:])
+	}
+
+	rootHash := storageTrie.Hash()
+
+	return rootHash, nil
+}
+
+// getValueInBytes returns the value of a key from SMT in bytes by traversing the SMT
+func (s *SMT) getValueInBytes(key int, address libcommon.Address) ([]byte, error) {
+	value := []byte{}
+
+	ethAddr := address.String()
+
+	kn, err := utils.Key(ethAddr, key)
+	if err != nil {
+		return nil, err
+	}
+
+	keyPath := kn.GetPath()
+
+	keyPathBytes := make([]byte, 0)
+	for _, k := range keyPath {
+		keyPathBytes = append(keyPathBytes, byte(k))
+	}
+
+	action := func(prefix []byte, k utils.NodeKey, v utils.NodeValue12) (bool, error) {
+		if !bytes.HasPrefix(keyPathBytes, prefix) {
+			return false, nil
+		}
+
+		if v.IsFinalNode() {
+			valHash := v.Get4to8()
+			v, err := s.Db.Get(*valHash)
+			if err != nil {
+				return false, err
+			}
+			vInBytes := utils.ArrayBigToScalar(utils.BigIntArrayFromNodeValue8(v.GetNodeValue8())).Bytes()
+
+			value = vInBytes
+			return false, nil
+		}
+
+		return true, nil
+	}
+
+	root, err := s.Db.GetLastRoot()
+	if err != nil {
+		return nil, err
+	}
+
+	err = s.Traverse(context.Background(), root, action)
+	if err != nil {
+		return nil, err
+	}
+
+	return value, nil
+}
+
+// getStorageMap returns the storage map of an address from the SMT
+func (s *SMT) getStorageMap(address libcommon.Address) (map[libcommon.Hash]libcommon.Hash, error) {
```

Comment:
This function traverses the entire trie every time. It could be optimized a bit by storing the map in memory after the first traversal.

Reply:
Done ✅